pkg/vsphere/vm/vm.go

```diff
@@ -310,17 +310,91 @@ func (vm *VirtualMachine) DeleteExceptDisks(ctx context.Context) (*object.Task,
 	return vm.Destroy(ctx)
 }
 
-// Unregister unregisters the VM
-func (vm *VirtualMachine) Unregister(ctx context.Context) error {
-	req := types.UnregisterVM{
-		This: vm.Reference(),
+func (vm *VirtualMachine) VMPathName(ctx context.Context) (string, error) {
+	var err error
+	var mvm mo.VirtualMachine
+
+	if err = vm.Properties(ctx, vm.Reference(), []string{"config.files"}, &mvm); err != nil {
+		log.Errorf("Unable to get vm config.files property: %s", err)
+		return "", err
+	}
+
+	return mvm.Config.Files.VmPathName, nil
+}
+
+func (vm *VirtualMachine) registerVM(ctx context.Context, path, name string,
+	vapp, pool, host *types.ManagedObjectReference, vmfolder *object.Folder) (*object.Task, error) {
+	log.Debugf("Register VM %s", name)
+
+	c := vm.Client.Client
+	if vapp == nil {
+		var hostObject *object.HostSystem
+		if host != nil {
+			hostObject = object.NewHostSystem(c, *host)
+		}
+		poolObject := object.NewResourcePool(c, *pool)
+		return vmfolder.RegisterVM(ctx, path, name, false, poolObject, hostObject)
 	}
 
-	_, err := methods.UnregisterVM(ctx, vm.Client.RoundTripper, &req)
+	req := types.RegisterChildVM_Task{
+		This: vapp.Reference(),
+		Path: path,
+		Host: host,
+	}
+
+	if name != "" {
+		req.Name = name
+	}
+
+	res, err := methods.RegisterChildVM_Task(ctx, vm.Client.Client, &req)
 	if err != nil {
+		return nil, err
+	}
+
+	return object.NewTask(vm.Client.Client, res.Returnval), nil
+}
+
+// FixInvalidState fix vm invalid state issue through unregister & register
+func (vm *VirtualMachine) FixInvalidState(ctx context.Context) error {
+	log.Debugf("Fix invalid state VM: %s", vm.Reference())
```

Comment:
A sanity check to see if the VM is indeed in an invalid state before attempting to unregister would be nice.